src/libsyntax/parse/parser.rs

```diff
@@ -3823,13 +3840,9 @@ impl<'a> Parser<'a> {
             token::BinOp(token::And) | token::AndAnd => {
                 // Parse &pat / &mut pat
                 self.expect_and()?;
+                self.maybe_disallow_lifetime();
                 let mutbl = self.parse_mutability();
-                if let token::Lifetime(ident) = self.token {
-                    let mut err = self.fatal(&format!("unexpected lifetime `{}` in pattern",
-                                                      ident));
-                    err.span_label(self.span, "unexpected lifetime");
-                    return Err(err);
-                }
+                self.maybe_disallow_lifetime();
```

Comment:
The only reason I added this second call is because the current check comes _after_ the `mut` parse, which is a weird place to have it, but I'm guessing not that uncommon of a mistake to write `foo(&mut 'a x)`...